### PR TITLE
Reference Implementation Of Multi-Wallet

### PR DESCRIPTION
**Pull request by neekolas**
## Summary

*DO NOT MERGE*

This is a reference implementation of the Multi-Wallet XIP, and is not meant to be used as production code. The test cases can be used as a reference of the soundness of the model. and a verification that known security exploits are handled by the model.

### CODE REVIEW

File: xmtp_id/src/associations/association_log.rs [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+        &self,
+        existing_state: Option<AssociationState>,
+    ) -> Result<AssociationState, AssociationError>;
+    fn hash(&self) -> String;
+}
+
+pub struct CreateXid {
+    pub nonce: u32,
+    pub account_address: String,
+    pub initial_association: AddAssociation,
```
Comment (richardhuaaa) [inline xmtp_id/src/associations/association_log.rs:43]: With the batching approach, we don't need this anymore - there will be a signature here right?

File: xmtp_id/src/associations/association_log.rs [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+    MissingExistingMember,
+    #[error("Legacy key is only allowed to be associated using a legacy signature with nonce 0")]
+    LegacySignatureReuse,
+    #[error("Signature not allowed for role {0:?} {1:?}")]
+    SignatureNotAllowed(EntityRole, SignatureKind),
+    #[error("Replay detected")]
+    Replay,
+}
+
+pub trait LogEntry {
```
[the remark this answers is not in the document]
Comment (richardhuaaa) [reply xmtp_id/src/associations/association_log.rs:32]: We might want to iron out the terminology between the protos and the implementation -
- member vs entity
- LogEntry vs IdentityUpdate vs IdentityAction
- AssociationEvent vs IdentityUpdate/Action

**Review by richardhuaaa — COMMENTED** [review]
Only a brief scan so far, but really like the structure. Are we aiming for this to be eventually committable? Just to adjust the level of detail I review this with

File: xmtp_id/src/associations/association_log.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (neekolas) [reply xmtp_id/src/associations/association_log.rs:32]: I'm working on a big renaming push.

Entity -> Member
LogEntry -> IdentityAction
AssociationEvent -> IdentityUpdate

File: xmtp_id/src/associations/association_log.rs (hunk 1 above)
Comment (neekolas) [reply xmtp_id/src/associations/association_log.rs:43]: That's right. That change came in after I wrote this, but I'll get it replaced. Should be straightforward.

File: xmtp_id/src/associations/entity.rs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+pub enum EntityRole {
+    Installation,
+    Address,
+}
+
+#[derive(Clone, Debug)]
+pub struct Entity {
+    pub role: EntityRole,
+    pub id: String,
+    pub added_by_entity: Option<String>,
```
[the remark this answers is not in the document]
Comment (neekolas) [reply xmtp_id/src/associations/entity.rs:11]: The only place it's used is to delete the children (if they are installations) during revocation.

**Comment by neekolas** [comment]
@richardhuaaa 

> Only a brief scan so far, but really like the structure. Are we aiming for this to be eventually committable? Just to adjust the level of detail I review this with

This PR won't be committable, but I have a new branch I'm working on that uses this as a base which I'm hoping will be.

**Pull request closed without merging**